src/control/controlindicatortimer.cpp

```diff
@@ -0,0 +1,32 @@
+#include "control/controlindicatortimer.h"
+
+#include "control/controlobject.h"
+#include "moc_controlindicatortimer.cpp"
+
+namespace mixxx {
+
+ControlIndicatorTimer::ControlIndicatorTimer(QObject* pParent)
+        : QObject(pParent),
+          m_pCOIndicator250millis(std::make_unique<ControlObject>(
+                  ConfigKey("[Master]", "indicator_250millis"))),
+          m_pCOIndicator500millis(std::make_unique<ControlObject>(
+                  ConfigKey("[Master]", "indicator_500millis"))),
+          m_toggleIndicator500millisOnNextTimeout(true) {
+    m_pCOIndicator250millis->setReadOnly();
+    m_pCOIndicator500millis->setReadOnly();
+    connect(&m_timer, &QTimer::timeout, this, &ControlIndicatorTimer::slotTimeout);
+    m_timer.start(250);
```

Comment:
The QTimer has special code that ensures to deliver the signal as close as possible to the given time.
That sounds appealing on the fist sight, unfortunately this introduces kind of priority for the connected events that interfere with the vsync code that has a hard deadline to met to avoid jerking. If a blink event happens during vsync, it is delayed together with the whole GUI until the next buffer is swapped. If there is pending waveform work to do, it is shiftet to the next buffer which is causing a jerk.

That was the original reason to schedule the repainting of blinking widgets with the vsync thread.

This is probably not relevant for QML widgets where the vsync is probably done later, but it still applies for CPU rendered widgets. Can we make the time source conditionally depending on the skin type?

I assume that we need some kind of vsync interval for any waveform solution.
So maybe it is best to provide an generic interface for such syncing.

What do you think?

Reply:
Would running the timer in its own thread solve this?

Reply:
> The QTimer has special code that ensures to deliver the signal as close as possible to the given time.

The default timer type is [`Qt::CoarseTimer`](https://doc.qt.io/qt-5/qt.html#TimerType-enum), that "try to keep accuracy within 5% of the desired interval". It's not a `Qt::PreciseTimer` which would try to keep millisecond accuracy.

> That sounds appealing on the fist sight, unfortunately this introduces kind of priority for the connected events that interfere with the vsync code that has a hard deadline to met to avoid jerking. If a blink event happens during vsync, it is delayed together with the whole GUI until the next buffer is swapped. If there is pending waveform work to do, it is shiftet to the next buffer which is causing a jerk.

If that was a problem, wouldn't that apply to *all* QTimers in the GUI thread? If so, why didn't we notice issues yet? We use QTimers often:

```
$ grep -Hrn "\bQTimer\b" ../src
../src/control/controlbehavior.h:3:#include <QTimer>
../src/control/controlbehavior.h:151:    QTimer* getTimer() {
../src/control/controlbehavior.h:153:            m_pushTimer.reset(new QTimer());
../src/control/controlbehavior.h:159:    QScopedPointer<QTimer> m_pushTimer;
../src/control/controlindicatortimer.h:4:#include <QTimer>
../src/control/controlindicatortimer.h:27:    QTimer m_timer;
../src/control/controlindicatortimer.cpp:17:    connect(&m_timer, &QTimer::timeout, this, &ControlIndicatorTimer::slotTimeout);
../src/widget/wtime.cpp:14:    m_pTimer = new QTimer(this);
../src/widget/wtime.cpp:25:    connect(m_pTimer, &QTimer::timeout, this, &WTime::refreshTime);
../src/widget/wsearchlineedit.h:6:#include <QTimer>
../src/widget/wsearchlineedit.h:95:    QTimer m_debouncingTimer;
../src/widget/wsearchlineedit.h:96:    QTimer m_saveTimer;
../src/widget/wtime.h:6:#include <QTimer>
../src/widget/wtime.h:25:    QTimer* m_pTimer;
../src/widget/wsearchlinee$ grep -HrnF QTimer ../src
../src/control/controlbehavior.h:3:#include <QTimer>
../src/control/controlbehavior.h:151:    QTimer* getTimer() {
../src/control/controlbehavior.h:153:            m_pushTimer.reset(new QTimer());
../src/control/controlbehavior.h:159:    QScopedPointer<QTimer> m_pushTimer;
../src/control/controlindicatortimer.h:4:#include <QTimer>
../src/control/controlindicatortimer.h:27:    QTimer m_timer;
../src/control/controlindicatortimer.cpp:17:    connect(&m_timer, &QTimer::timeout, this, &ControlIndicatorTimer::slotTimeout);
../src/vinylcontrol/vinylcontrolmanager.cpp:169:void VinylControlManager::timerEvent(QTimerEvent*) {
../src/vinylcontrol/vinylcontrolsignalwidget.h:13:#include <QTimerEvent>
../src/vinylcontrol/vinylcontrolmanager.h:5:#include <QTimerEvent>
../src/vinylcontrol/vinylcontrolmanager.h:43:    void timerEvent(QTimerEvent* pEvent);
../src/mixer/playerinfo.h:8:#include <QTimerEvent>
../src/mixer/playerinfo.h:49:    void timerEvent(QTimerEvent* pTimerEvent) override;
../src/mixer/playerinfo.cpp:112:void PlayerInfo::timerEvent(QTimerEvent* pTimerEvent) {
../src/widget/wtime.cpp:14:    m_pTimer = new QTimer(this);
../src/widget/wtime.cpp:25:    connect(m_pTimer, &QTimer::timeout, this, &WTime::refreshTime);
../src/widget/wsearchlineedit.h:6:#include <QTimer>
../src/widget/wsearchlineedit.h:95:    QTimer m_debouncingTimer;
../src/widget/wsearchlineedit.h:96:    QTimer m_saveTimer;
../src/widget/wtime.h:6:#include <QTimer>
../src/widget/wtime.h:25:    QTimer* m_pTimer;
../src/widget/wsearchlineedit.cpp:116:            &QTimer::timeout,
../src/widget/wsearchlineedit.cpp:120:            &QTimer::timeout,
../src/widget/wcoverart.h:7:#include <QTimer>
../src/widget/wcoverart.h:76:    QTimer m_clickTimer;
../src/widget/wpushbutton.h:8:#include <QTimer>
../src/widget/wpushbutton.h:110:    QTimer m_clickTimer;
../src/widget/wlibrarysidebar.cpp:110:void WLibrarySidebar::timerEvent(QTimerEvent *event) {
../src/widget/wlibrarysidebar.h:10:#include <QTimerEvent>
../src/widget/wlibrarysidebar.h:26:    void timerEvent(QTimerEvent* event) override;
../src/library/dlgcoverartfullsize.h:5:#include <QTimer>
../src/library/dlgcoverartfullsize.h:50:    QTimer m_clickTimer;
../src/library/sidebarmodel.h:6:#include <QTimer>
../src/library/sidebarmodel.h:90:    QTimer* const m_pressedUntilClickedTimer;
../src/library/sidebarmodel.cpp:34:          m_pressedUntilClickedTimer(new QTimer(this)) {
../src/library/sidebarmodel.cpp:37:            &QTimer::timeout,
../src/waveform/guitick.cpp:1:#include <QTimer>
../src/dialog/dlgdevelopertools.cpp:92:void DlgDeveloperTools::timerEvent(QTimerEvent* pEvent) {
../src/dialog/dlgdevelopertools.h:6:#include <QTimerEvent>
../src/dialog/dlgdevelopertools.h:20:    void timerEvent(QTimerEvent* pTimerEvent) override;
../src/util/timer.h:112:// A timer that provides a similar API to QTimer but uses render events from the
../src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp:471:void ControllerScriptInterfaceLegacy::timerEvent(QTimerEvent* event) {
../src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h:64:    virtual void timerEvent(QTimerEvent* event);
../src/controllers/controllermanager.h:4:#include <QTimer>
../src/controllers/controllermanager.h:77:    QTimer m_pollTimer;
../src/controllers/dlgcontrollerlearning.h:6:#include <QTimer>
../src/controllers/dlgcontrollerlearning.h:83:    QTimer m_firstMessageTimer;
../src/controllers/dlgcontrollerlearning.h:84:    QTimer m_lastMessageTimer;
../src/controllers/dlgcontrollerlearning.cpp:137:    connect(&m_lastMessageTimer, &QTimer::timeout, this, &DlgControllerLearning::slotTimerExpired);
../src/controllers/dlgcontrollerlearning.cpp:142:            &QTimer::timeout,
../src/controllers/controller.h:4:#include <QTimerEvent>
../src/controllers/controllermanager.cpp:101:    connect(&m_pollTimer, &QTimer::timeout, this, &ControllerManager::pollDevices);
../src/network/jsonwebtask.cpp:6:#include <QTimerEvent>
../src/network/webtask.cpp:4:#include <QTimerEvent>
../src/network/webtask.cpp:300:void WebTask::timerEvent(QTimerEvent* event) {
../src/network/webtask.h:146:    void timerEvent(QTimerEvent* event) final;
```dit.cpp:116:            &QTimer::timeout,
../src/widget/wsearchlineedit.cpp:120:            &QTimer::timeout,
../src/widget/wcoverart.h:7:#include <QTimer>
../src/widget/wcoverart.h:76:    QTimer m_clickTimer;
../src/widget/wpushbutton.h:8:#include <QTimer>
../src/widget/wpushbutton.h:110:    QTimer m_clickTimer;
../src/library/dlgcoverartfullsize.h:5:#include <QTimer>
../src/library/dlgcoverartfullsize.h:50:    QTimer m_clickTimer;
../src/library/sidebarmodel.h:6:#include <QTimer>
../src/library/sidebarmodel.h:90:    QTimer* const m_pressedUntilClickedTimer;
../src/library/sidebarmodel.cpp:34:          m_pressedUntilClickedTimer(new QTimer(this)) {
../src/library/sidebarmodel.cpp:37:            &QTimer::timeout,
../src/waveform/guitick.cpp:1:#include <QTimer>
../src/util/timer.h:112:// A timer that provides a similar API to QTimer but uses render events from the
../src/controllers/controllermanager.h:4:#include <QTimer>
../src/controllers/controllermanager.h:77:    QTimer m_pollTimer;
../src/controllers/dlgcontrollerlearning.h:6:#include <QTimer>
../src/controllers/dlgcontrollerlearning.h:83:    QTimer m_firstMessageTimer;
../src/controllers/dlgcontrollerlearning.h:84:    QTimer m_lastMessageTimer;
../src/controllers/dlgcontrollerlearning.cpp:137:    connect(&m_lastMessageTimer, &QTimer::timeout, this, &DlgControllerLearning::slotTimerExpired);
../src/controllers/dlgcontrollerlearning.cpp:142:            &QTimer::timeout,
../src/controllers/controllermanager.cpp:101:    connect(&m_pollTimer, &QTimer::timeout, this, &ControllerManager::pollDevices);
```

Reply:
I think accuracy within 5% should be fine for blinking these COs.

Reply:
> If a blink event happens during vsync, it is delayed together with the whole GUI until the next buffer is swapped.

This seems to be a misunderstanding as @Holzhaus explained above. Testing this branch, I do not see any regression in waveform jerking and the indicator controls continue to work.

Reply:
I switched between this branch and main several times playing 2 decks with 2 decks paused using the Pioneer cue mode which has (annoyingly) frequent blinking and did not notice a difference in waveform scrolling.